core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -64,18 +66,54 @@
 
 	private static final Logger logger = LoggerFactory.getLogger(Transaction.class);
 
+	/**
+	 * The period during the {@link #close()} method between waiting for
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)}, logging that the transaction has not
+	 * completed, and trying again up to {@link #MAX_CLOSE_WAIT_PERIODS} attempts.
+	 */
+	private static final int EXECUTOR_SHUTDOWN_WAIT_PERIOD = 10000;
+
+	/**
+	 * The maximum number of times during the {@link #close()} method that
+	 * {@link ExecutorService#awaitTermination(long, TimeUnit)} will be called before giving up and calling
+	 * {@link ExecutorService#shutdownNow()}.
+	 */
+	private static final int MAX_CLOSE_WAIT_PERIODS = 3;
+
+	/**
+	 * Set to true when entering the {@link #close()} method for the first time, to ensure that only a single
+	 * thread executes the close operations.
+	 */
 	private final AtomicBoolean isClosed = new AtomicBoolean(false);
 
+	/**
+	 * Set to true when the {@link #close()} method is about to complete for the first invocation.
+	 */
+	private final AtomicBoolean closeCompleted = new AtomicBoolean(false);
+
 	private final UUID id;
 
 	private final Repository rep;
 
 	private final RepositoryConnection txnConnection;
 
+	/**
+	 * The {@link ExecutorService} that performs all of the operations related to this Transaction.
+	 */
 	private final ExecutorService executor = Executors.newSingleThreadExecutor(
 			new ThreadFactoryBuilder().setNameFormat("rdf4j-transaction-%d").build());
 
-	private final List<Future<?>> futures = new ArrayList<>();
+	/**
+	 * An append-only {@link Queue} that tracks the state of all of the futures involved in this Transaction.
+	 */
+	private final Queue<Future<?>> futures = new LinkedBlockingQueue<>();
+
+	/**
+	 * This {@link Semaphore} is used to ensure that only a single task is executed by the ExecutorService,
+	 * and that these tasks are not run while {@link #hasActiveOperations()} is determining whether there are
+	 * outstanding tasks open for this Transaction.
+	 */
+	private final Semaphore semaphore = new Semaphore(1);
```

Comment:
Yes. My assumption, based on looking through code lines related to #684 was that high accuracy on the hasActiveOperations method is required to make sure the Transaction is not closed prematurely, which may have in the past caused BackgroundGraphResult threads to be isolated by broken close methods throughout the codebase wrt exceptions being thrown while close was proceeding.

The use of an internal single threaded ExecutorService gives ordering guarantees only if the user had already given operations in order (which they can't given the asynchronous closing of transactions in ActiveTransactionRegistry). These changes make both the threads inside of the ExecutorService and the ActiveTransactionRegistry get (almost) atomic snapshots of the state, so when the close blocks on the semaphore it also blocks calls to hasActiveOperations until it is ready to return. If hasActiveOperations begins and gets both the lock (so new operations cannot be submitted), and the semaphore, (so existing operations are either blocked on starting, or have only just finished), then it receives as accurate a state as possible while it is iterating. The only case then is a safe case with respect to cache removal, that of "hasActiveOperations" returning true because a Callable has just released the semaphore but the ExecutorService has not yet set isDone to true.

However, that may be entirely irrelevant and the semaphore can be easily removed and hasActiveOperations would then be independent again of the state of running tasks.